src/compiler/dataStructures.ts

```diff
@@ -0,0 +1,768 @@
+    export function keysArray<V>(map: SMap<V>): string[] {
+        const keys: string[] = [];
+        forEachKeyInMap(map, key => keys.push(key));
+        return keys;
+    }
+
+    /** Array of all values in a map. */
+    export function valuesArray<K, V>(map: SMap<V>): V[] {
+        const values: V[] = [];
+        map.forEach(value => values.push(value));
+        return values;
+    }
+
+    /** Set `map.get(key) === value`, and return `value`. */
+    export function setAndReturn<K, V>(map: MapCommon<K, V>, key: K, value: V): V {
+        map.set(key, value);
+        return value;
+    }
+
+    /** Like `getOrUpdateMap` but does not return a value. */
+    export function setButDontOverride<K, V>(map: MapCommon<K, V>, key: K, getValue: () => V): void {
+        if (!map.has(key)) {
+            map.set(key, getValue());
+        }
+    }
+
+    /** Sets `map.get(key) === getValue()` unless `map.get(key)` was already defined. */
+    export function getOrUpdateMap<K, V>(map: MapCommon<K, V>, key: K, getValue: () => V): V {
+        const value = map.get(key);
+        if (value === undefined) {
+            const value = getValue();
+            map.set(key, value);
+            return value;
+        }
+        else {
+            return value;
+        }
+    }
+
+    /** True iff the predicate is true for some entry in the map. */
+    export function someInMap<V>(map: SMap<V>, predicate: (value: V, key: string) => boolean): boolean {
+        return !!findInMap(map, predicate);
+    }
+
+    /** True iff the predicate is true for every entry in the map. */
+    export function allInMap<V>(map: SMap<V>, predicate: (value: V, key: string) => boolean): boolean {
+        return !findInMap(map, (value, key) => !predicate(value, key));
+    }
+
+    /** Array of every *defined* result of `mapAndFilter(value, key)`, called on each entry in the map. */
+    export function mapAndFilterMap<V, U>(map: SMap<V>, mapAndFilter: (value: V, key: string) => U | undefined): U[] {
+        const result: U[] = [];
+        map.forEach((value, key) => {
+            const entry = mapAndFilter(value, key);
+            if (entry !== undefined) {
+                result.push(entry);
+            }
+        });
+        return result;
+    }
+
+    /**
+     * Adds the value to an array of values associated with the key, and return the array.
+     * Creates the array if it does not already exist.
+     */
+    export function multiMapAdd<K, V>(map: MapCommon<K, V[]>, key: K, value: V): V[] {
+        const values = map.get(key);
+        if (values) {
+            values.push(value);
+            return values;
+        }
+        else {
+            return setAndReturn(map, key, [value]);
+        }
+    }
+
+    /** Creates a map with the given keys and values for each entry in `inputs`. */
+    export function createMapFromArray<A, K, V>(inputs: A[], getKey: (element: A) => string, getValue: (element: A) => V): SMap<V> {

```

Comment:
I don't think K is used here ? also since it always return `SMap` it will be appropriate to rename the function to `createStringMapFromArray` to be more descriptive